pageserver/src/bin/pageserver.rs

```diff
@@ -277,12 +277,30 @@ fn start_pageserver(conf: &'static PageServerConf) -> anyhow::Result<()> {
         }
     };
 
-    let remote_storage = conf
-        .remote_storage_config
-        .as_ref()
-        .map(GenericRemoteStorage::from_config)
-        .transpose()
-        .context("Failed to init generic remote storage")?;
+    let remote_storage = if let Some(remote_storage_conf) = &conf.remote_storage_config {
+        let remote_storage = GenericRemoteStorage::from_config(remote_storage_conf)
+            .context("Failed to init generic remote storage")?;
+
+        if conf.test_remote_failures > 0 {
+            if cfg!(feature = "testing") {
+                info!(
+                    "Simulating remote failures for first {} attempts of each op",
+                    conf.test_remote_failures
+                );
+                Some(GenericRemoteStorage::unreliable_wrapper(
+                    remote_storage,
+                    conf.test_remote_failures,
+                ))
+            } else {
+                anyhow::bail!("test_remote_failures option is not available because pageserver was compiled without the 'testing' feature");
+            }
+        } else {
+            //Some(remote_storage)
```

Comment:
Some leftovers, I think this should be returned back.
Super odd that the tests did not fail...

I also agree with the comment above, that the init code is odd.
I think, all related remote storage code should be barred behind `#[cfg(testing)]` feature ideally, to cut it away statically.
If we follow with the special error type for the simulated failure and will add a new variant into `DownloadError`, that variant could be also compiled out, IIRC, same as the config-related code.

Reply:
>Some leftovers, I think this should be returned back.

Oops, yes.

> Super odd that the tests did not fail...

Shows that we have retries in all the right places now :-).

Reply:
Fixed that leftover, and rebased this over PR #3145.

Reply:
> I think, all related remote storage code should be barred behind #[cfg(testing)] feature ideally, to cut it away statically.
If we follow with the special error type for the simulated failure and will add a new variant into DownloadError, that variant could be also compiled out, IIRC, same as the config-related code.

I know we disagree on this, but I'd rather go in the opposite direction, per https://github.com/neondatabase/neon/pull/3030. Conditionally-compiled code is annoying, as you need to compile the code with all combinations of the feature flags to make sure that all the variants compile. Some dead code in production binaries doesn't do any harm.